### PR TITLE
[BEAM-2986] Support for reading avro GenericRecords from BigQuery

### PR DESCRIPTION
**Pull request by steveniemitz**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

This is an initial pass at support for reading `GenericRecord`s from BigQuery (rather than JSON TableRows).

A couple points of discussion:
- `SerializableFunction<TableSchema, SerializableFunction<GenericRecord, T>>` is a disgusting generic type, should this be refactored to be simply `ParseFnFactory<T>` or similar?
- I removed `BigQueryJsonReader` and all references to it.  A non-split `createReader` now follows the same code-path as split readers.  I'm not sure if there's a better way to do this than introducing an aggregating reader like I did with `MergedBoundedReader`?  I'm not sure under which conditions each method is called, but this seemed like the best compromise.
- I followed the AvroIO convention of `readRecords().withParseFn(...)`, however it seems like it would make a better API to make `readRecords` take the `parseFn` directly, so it's impossible to even attempt to build a reader without a parseFn (that would then fail at `expand()` time.  Is there a reason for the design of AvroIO like that?

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 1]
```diff
@@ -319,149 +328,193 @@ public static Read read() {
...
+    return new AutoValue_BigQueryIO_ReadGenericRecords.Builder<T>()
+        .setValidate(true)
+        .setWithTemplateCompatibility(false)
+        .setBigQueryServices(new BigQueryServicesImpl())
+        .build();
+  }
 
-    abstract Boolean getWithTemplateCompatibility();
+  private static class TableSchemaFunction
+      implements Serializable, Function<String, TableSchema> {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This should use a SerializableFunction

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 2]
```diff
@@ -319,149 +328,193 @@ public static Read read() {
...
-     * href="https://cloud.google.com/bigquery/docs/reference/v2/jobs">Jobs documentation</a> for
-     * more information. To disable flattening, use {@link BigQueryIO.Read#withoutResultFlattening}.
-     *
-     * <p>By default, the query will use BigQuery's legacy SQL dialect. To use the BigQuery Standard
-     * SQL dialect, use {@link BigQueryIO.Read#usingStandardSql}.
-     */
-    public Read fromQuery(String query) {
-      return fromQuery(StaticValueProvider.of(query));
+    @Override
+    Coder<TableRow> getCoder() {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: I would prefer to use regular setters in the read() factory method, rather than method overrides, to set property values.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 3]
```diff
@@ -319,149 +328,193 @@ public static Read read() {
...
-      return toBuilder().setWithTemplateCompatibility(true).build();
+    public ReadGenericRecords<T> withCoder(Coder<T> coder) {
+      return toBuilder().setCoder(coder).build();
     }
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  abstract static class ReadBuilderBase<ReadT, SelfT extends ReadBuilderBase<ReadT, SelfT>> {
+    abstract SelfT setJsonTableRef(ValueProvider<String> jsonTableRef);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: I'm impressed that AutoValue can handle this!

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 4]
```diff
@@ -577,21 +633,21 @@ public String apply(String input) {
...
-                                    source.getSchema(c.getPipelineOptions()));
-                            c.output(tableSchemaTag, schema);
-                            List<ResourceId> files = source.extractFiles(c.getPipelineOptions());
-                            for (ResourceId file : files) {
-                              c.output(file.toString());
-                            }
-                          }
-                        })
+                    new DoFn<String, String>() {
+                      @ProcessElement
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: It would be good to put the formatting changes into a separate commit (first commit in the chain ideally) to avoid polluting the diff.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java [hunk 5]
```diff
@@ -128,13 +133,23 @@ protected TableSchema getSchema(PipelineOptions options) throws Exception {
   protected abstract void cleanupTempResource(BigQueryOptions bqOptions) throws Exception;
 
   @Override
+  public BoundedReader<T> createReader(PipelineOptions options) throws IOException {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: Why do we need this and MergedBoundedReader? Sources are always split by the runner before reading, so you don't need to implement createReader() on a source that is not the result of a split() call.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java [hunk 6]
```diff
@@ -167,69 +182,70 @@ public void validate() {
...
-          private Supplier<TableSchema> schema = Suppliers.memoize(
-              Suppliers.compose(new TableSchemaFunction(), Suppliers.ofInstance(jsonSchema)));
-
-          @Override
-          public TableRow apply(GenericRecord input) {
-            return BigQueryAvroUtils.convertGenericRecordToTableRow(input, schema.get());
-          }};
 
-    List<BoundedSource<TableRow>> avroSources = Lists.newArrayList();
+    SerializableFunction<GenericRecord, T> function = parseFnFactory.apply(tableSchema);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: I'm wondering whether tableSchema is needed as a parameter at all. GenericRecord already has a Schema (getSchema()) - it's not the same as the TableSchema, but it probably contains all the same relevant information. I suspect in most cases users wouldn't use the schema at all because they know what fields they are querying for; and in cases where they don't, I think GenericRecord.getSchema() will be sufficient.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 3 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: lol it took me like an hour to coax AutoValue into getting this right.  Overriding `toBuilder()` seemed like the magic incantation to get it to compile.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java (hunk 5 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: I'm not sure if we do need it.  Is there actually a guarantee somewhere that says BoundedSource will never have `createReader` called on it?  The API design is confusing here, and doing a find-references on `createReader` do show it being called.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java (hunk 6 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: Yeah, I could spend some time trying to get this to work.  I was trying to avoid having to re-implement the whole existing avro -> TableRow process here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 1 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: can do, I just moved this out of BigQuerySourceBase though :P

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 2 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: 👍 makes sense.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 4 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: ugh yeah, sadly a lot of the reformatting was intellij imposing its coding style on the code.  I can try to un-format it.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 1 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: jk, its not a SerializableFunction apparently because it needs to extend guava's `Function` for `Suppliers.compse` below.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 2 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: ah upon trying to refactor this I realized why these are like they are: AutoValue has issues generating the code for `Read` since the parameter is "reified" into TableRow.

```
[ERROR] /Users/sniemitz/dev/beam/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java:[395,24] Parameter type org.apache.beam.sdk.coders.Coder<com.google.api.services.bigquery.model.TableRow> of setter method should be org.apache.beam.sdk.coders.Coder<T> to match getter org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Read.getCoder
```

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java (hunk 6 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: So, please correct me if I'm wrong, but I don't think the avro scheme has enough information to maintain full backwards compatibility with the current implementation.

The big edge case I can think of right now is that timestamps in BigQuery are represented as `long`s in avro, and without the BigQuery schema, it's impossible to re-format them as dates and not just numbers (see BigQueryAvroUtils.java#204.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 2 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Hmm, I think it may be better to just bite the bullet and have some boilerplate (separate builder classes for Read and ReadGenericRecords, perhaps the former delegating to the latter like in TextIO.Write): if today it required hackery to get to compile, one day it might require hackery for somebody to fix it after an innocuous change that will somehow be "too much" for AutoValue.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java (hunk 5 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: Yeah, it is not documented, but it is true in practice for all runners, and it is the reason why removing the json reading codepaths with no replacement is safe (see my PR).

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java (hunk 6 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: Hm. I really don't want to subject users to having to implement a nested `SerializableFunction`. One slightly less ugly thing that comes to mind is suggest to create a helper `class BigQueryAvroRecord implements GenericRecord {TableSchema getBigQuerySchema();}` (which delegates the other methods of GenericRecord to the actual GenericRecord) and make the user pass a `SerializableFunction<BigQueryAvroRecord, T>`? Any obvious downsides to this? (I'm not  yet confident that this is actually good)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 2 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Yeah, at this point it probably makes sense to just ditch the AutoValue stuff here, I don't know if its adding much value (lol no pun intended).

I'll give that a try.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java (hunk 5 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: ok cool.  I'll rebase once that other PR lands.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java (hunk 6 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: I'm very torn on this in general.  On one hand having a higher-order function like this isn't very abnormal in other languages.  On the other hand Java makes this stupidly verbose, ugly, and hard to implement. 
Your proposal makes sense, but I'm not how sure how much I love squishing the BQ schema and avro record together like that.  If anything I'd prefer composition over inheritance, and have something like:
```
class SchemaAndRecord {
   private final GenericRecord record;
   private final TableSchema schema;
 
  // lots of java boilerplate.
}
```
But to me, this object model implies that the schema might change for each call and isn't constant.

Alternatively we could do something like `SerializableFunction2<TableSchema, GenericRecord, T>`, or like I suggested above something more specific to BigQuery like:
```
class ParseFn<T> extends Serializable { 
  T apply(TableSchema tableSchema, GenericRecord record) {...}
}
```
Again though this has similar problems as above.

Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: Composition is fine too. I agree that it's funky that it implies schema might change, but this is already true of the GenericRecord's `getSchema()` so I think adding another schema argument that doesn't really change doesn't make this worse.

To make this seem less horrible, consider that one day we'll probably add BigQueryIO.readAll() that reads a PCollection of table names or queries. Then the schema might actually be different, even though for any given schema there will be a lot of records with that schema.

Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java]: I'm sold :)

I'll cook something up.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 2 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Based on the suggestion below of `Read` simply wrapping `ReadGenericRecords`, this becomes much simpler since I no longer need a base class for the PTransform and Builders.

**Comment by steveniemitz** [comment]
updated with comments, ptal @jkff :)

**Comment by jkff** [comment]
Could you rebase one more time please? I submitted the PR with removal of JSON reading codepaths.

**Comment by steveniemitz** [comment]
Yep, just saw the email, `mvn clean verify` in progress right now

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 7]
```diff
@@ -312,105 +315,141 @@ public TableRow apply(TableRow input) {
...
+      this.inner = inner;
     }
 
-    /** Ensures that methods of the from() / fromQuery() family are called at most once. */
-    private void ensureFromNotCalledYet() {
-      checkState(
-          getJsonTableRef() == null && getQuery() == null, "from() or fromQuery() already called");
+    @Override
+    public PCollection<TableRow> expand(PBegin input) {
+      return inner.expand(input);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Better do input.apply(inner). expand must not be called directly by anybody except a runner. I think you also don't need populateDisplayData, validate, or others - inner will do this for you automatically.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 8]
```diff
@@ -312,105 +315,141 @@ public TableRow apply(TableRow input) {
...
+      return new Read(this.inner.withTestServices(testServices));
+    }
+
+    ///////////////////////////////////////////////////////////////////
+
+    /**
+     * Returns the table to read, or {@code null} if reading from a query instead.
+     */
+    @Nullable
+    public ValueProvider<TableReference> getTableProvider() {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Was it previously public?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 9]
```diff
@@ -312,105 +315,141 @@ public TableRow apply(TableRow input) {
...
+      return this.inner.getTableProvider();
+    }
+
+    /** Returns the table to read, or {@code null} if reading from a query instead. */
+    @Nullable
+    public TableReference getTable() {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Ditto

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 10]
```diff
@@ -442,26 +481,95 @@ public Read usingStandardSql() {
...
     public Read withTemplateCompatibility() {
-      return toBuilder().setWithTemplateCompatibility(true).build();
+      return new Read(this.inner.withTemplateCompatibility());
+    }
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Implementation of {@link BigQueryIO#readRecords(SerializableFunction)} )}.
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Trailing ) }

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 11]
```diff
@@ -442,26 +481,95 @@ public Read usingStandardSql() {
...
+    Coder<T> inferCoder(CoderRegistry coderRegistry) {
+      if (getCoder() != null) {
+        return getCoder();
+      }
+
+      TypeDescriptor<T> descriptor = TypeDescriptors.outputOf(getParseFn());
+
+      String message =
+          "Unable to infer coder for output of parseFn. Specify it explicitly using withCoder().";
+      checkArgument(descriptor != null, message);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This was copied from AvroIO but it's buggy there - actually descriptor can never be null, you can omit this check.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 12]
```diff
@@ -610,31 +720,31 @@ public String apply(String s) {
...
+                                BoundedSource.BoundedReader<T> reader =
                                     avroSource.createReader(c.getPipelineOptions());
                                 for (boolean more = reader.start(); more; more = reader.advance()) {
                                   c.output(reader.getCurrent());
                                 }
                               }
                             })
                         .withSideInputs(schemaView, jobIdTokenView))
-                        .setCoder(TableRowJsonCoder.of());
+                        .setCoder(inferCoder(input.getPipeline().getCoderRegistry()));
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Shouldn't this be simply "coder"?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 13]
```diff
@@ -697,6 +813,103 @@ public TableReference getTable() {
       ValueProvider<TableReference> provider = getTableProvider();
       return provider == null ? null : provider.get();
     }
+
+    public ReadGenericRecords<T> withCoder(Coder<T> coder) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Document this one too

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 14]
```diff
@@ -697,6 +813,103 @@ public TableReference getTable() {
...
+      return toBuilder().setCoder(coder).build();
+    }
+
+    /**
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: I'd recommend to avoid the duplication of javadocs and document all of these methods as "Like {@link Read#sameMethod}", to reduce the burden for future maintainers.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 15]
```diff
@@ -127,7 +130,7 @@
  *
  * <h3>Reading</h3>
  *
- * <p>To read from a BigQuery table, apply a {@link BigQueryIO.Read} transformation. This produces a
+ * <p>To read from a BigQuery table, apply a {@link Read} transformation. This produces a
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: The class-level comment should mention the new methods and ideally have a short example.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 7 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: fwiw I was just copying what `TextIO.Write` does here.  Also removing the delegation to populateDisplayData, etc breaks some unit tests.  Should I still drop the overrides and change the `expand` to what you suggested?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 8 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: [it was](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L680)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 9 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: [ditto x2](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L686)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 10 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: 👍

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 11 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: The function is annotated with `@Nullable`, is that a mistake?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 12 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: 👍

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 7 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Hm, I guess it doesn't hurt to keep the populateDisplayData method; but you definitely should change .expand() to .apply(). And validate() is definitely unnecessary.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 11 above)
Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Hm which code path are you referring to? I don't see how that can happen, looking at the code currently. It delegates to extractFromTypeParameters which has only 1 return statement, and it returns a type argument of a parameterized type, which can't be null.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 7 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: 👍 sounds good.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 11 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: I see `outputOf` -> `TypeDescriptors.extractFromTypeParameters(type, supertype, extractor)` which has:

```
if (res.hasUnresolvedParameters()) {
      return null;
    } else {
      return (TypeDescriptor<V>) res;
    }
```

https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/TypeDescriptors.java#L344

Comment (jkff) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Oh, sorry, you're right. I was looking at code in my IDE, looking at a branch where I had changed this, and the branch had been around for so long that I forgot whether this part was submitted or not :-|

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 14 above)
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: 👍

**Comment by steveniemitz** [comment]
Its hard to do an apples-apples comparison here because the model is so different between `read` and `readGenericRecords`, but I did a trivial test of reading ~100 million rows from a BQ table and summing one field of them.  I used 5 dataflow workers (default configuration) w/o autoscaling enabled.

The table was 96,470,755 rows, and BQ reported it to be 19.4 GB.

Read Method | Wall time* | Throughput** | Bytes processed*** | size increase/(decrease)
--------------|-----------|-------------|--------------------|----------------------
`read()`          | 15:39        | ~140,000    | 163.61 GB                | 8.5x
`readGenericRecords()` | 7:49 | ~700,000 | 1.71 GB               | (11)x

\* Wall time = time from job starting to the last step completing.
** In rows per second, measured from the input to the `Sum` PTransform.
*** Bytes processed measured at the input to the step after the read transform

The two flows are here: https://gist.github.com/steveniemitz/c6de0b7f6dbfbee53e812956a4f5476a

**Comment by jkff** [comment]
Wonderful, thanks! Merging.

**Comment by steveniemitz** [comment]
Thanks for the help on this one!

**Comment by jkff** [comment]
Merged. I took the liberty of doing a bit of a documentation cleanup and a couple of renames, to encourage people to use the new method because it's so much more performant. Specifically:

* `read()` is deprecated
* Added `readTableRows()` as a synonym for `read()` but that is less tempting to use as a "default"
* Renamed `readGenericRecords(SerializableFunction)` to `read(SerializableFunction)` so that by default people feel encouraged to use this one.

**Comment by steveniemitz** [comment]
So cool!  Thats a great idea to drive usage.  Thanks again!
